src/main/java/io/confluent/idesidecar/websocket/resources/WebsocketEndpoint.java

```diff
@@ -265,16 +197,101 @@ public void onMessage(String messageString, Session senderSession) {
+   * @param messageString the JSON string message.
+   */
+  private void handleHelloMessage(Session session, String messageString) {
+    // Regardless of if the message was good or bad, we remove the session from purgatory
+    // upon receipt of the first message.
+    purgatorySessions.remove(session);
+
+    // deserializeMessage() will close the connection if the message is invalid and
+    // not a HELLO_WORKSPACE message
+    var message = deserializeMessage(session, messageString, MessageType.WORKSPACE_HELLO);
+    if (message == null) {
+      Log.errorf(
+          "handleHelloMessage: Invalid message from purgatory session %s. Closed session.",
+          session.getId()
+      );
+    }
+
+    // Message is valid and is a HELLO_WORKSPACE message with a HelloBody.
+    var workspaceId = ((HelloBody) message.body()).workspaceId();
+
+    // Ensure that the header origin value corresponds to the
+    // workspace id in the body.
+    if (!validateOriginator(session, message, workspaceId)) {
+      sendErrorAndCloseSession(
+          session,
+          message.headers().id(),
+          "handleHelloMessage: Workspace id %s does not match originator value. Closing session.",
+          workspaceId
+      );
+      return;
+    }
+
+    // Cross-reference the workspace id against the known workspaces.
+
+    if (!knownWorkspacesBean.isKnownWorkspace(Long.valueOf(workspaceId))) {
+      sendErrorAndCloseSession(
+          session,
+          message.headers().id(),
+          "handleHelloMessage: Unknown workspace id %d. Closing session.",
+          workspaceId
+      );
+      return;
+    }
+
+    // Should be unique across all workspaces in session map.
+    for(var ws : sessions.values()) {
+      if (ws.processId() == workspaceId) {
+        sendErrorAndCloseSession(
+            session,
+            message.headers().id(),
+            "Workspace id %s already connected. Closing session.",
+            workspaceId
+        );
+        return;
+      }
+    }
+
+    // All good! Add the workspace to the sessions map and announce the count change.
+    sessions.put(session, new WorkspaceSession(workspaceId));
+    Log.infof("Workspace %d authorized and added to sessions.", workspaceId);
+
+    // Announce to all other workspaces the list has changed
+    try {
+      broadcastWorkspacesChanged();
+    } catch (IOException e) {
+      Log.errorf("Failed to broadcast workspace added message: %s", e.getMessage());
+    }
+  }
+
   @OnError
   public void onError(Session session, Throwable throwable) throws IOException {
     // May or may not actually remove -- if had not yet been authorized, it won't be in the map.
     // (but will definitely not be in the map after this statement.)
     var existingSession = sessions.remove(session);
+    if (existingSession == null) {
+      // Must have been in purgatory and had an error before saying hello.
+      purgatorySessions.remove(session);
+    }
+
```

Comment:
Using a single collection would also improve this logic and guarantee it either happens or does not happen atomically. 

Sorry for pointing you toward a separate "purgatory" collection. :-(

Reply:
Done.